public/docs/ts/latest/guide/style-guide.jade

```diff
@@ -0,0 +1,672 @@
+      }
+
+      // private functions
+      private _show() {
+        console.log(this.message);
+        this._toastElement.style.opacity = 1;
+        this._toastElement.style.zIndex = 9999;
+
+        window.setTimeout(() => this._hide(), 2500);
+      }
+
+      private _hide() {
+        this._toastElement.style.opacity = 0;
+        window.setTimeout(() => this._toastElement.style.zIndex = 0, 400);
+      }
+    }
+    ```
+
+<a href="#toc">Back to top</a>
+
+:marked
+  ### Defer Logic to Services
+  <a id="035"></a>
+  #### Style 035
+
+    - Defer logic in a component by delegating to services.
+
+    - Define a component for a view, and try not to reuse the component for other views. Instead, move reusable logic to factories and keep the component simple and focused on its view.
+
+      **Why?** Logic may be reused by multiple components when placed within a service and exposed via a function.
+
+      **Why?** Logic in a service can more easily be isolated in a unit test, while the calling logic in the component can be easily mocked.
+
+      **Why?** Removes dependencies and hides implementation details from the component.
+
+      **Why?** Keeps the component slim, trim, and focused.
+
+  ```typescript
+  // avoid
+  export class SessionListComponent implements OnInit {
+    sessions: Session[];
+
+    constructor(private _http: Http) { }
+
+    getSessions() {
+      this.sessions = [];
+      this._http.get(sessionsUrl)
+        .map((response: Response) => <Session[]>response.json().data)
+        .catch(this._exceptionService.catchBadResponse)
+        .finally(() => this._spinnerService.hide())
+        .subscribe(sessions => this.sessions = sessions);
+    }
+    
+
+    ngOnInit() {
+      this.getSessions();
+    }
+  }
+  ```
+    
+  ```typescript
+  // recommended
+  export class SessionListComponent implements OnInit {
+    sessions: Session[];
+
+    constructor(private _sessionService: SessionService) { }
+
+    getSessions() {
+      this.sessions = [];
+      this._sessionService.getSessions()
+        .subscribe(sessions => this.sessions = sessions);
+    }
+
+    ngOnInit() {
+      this.getSessions();
+    }
+  }
+  ```
+
+<a href="#toc">Back to top</a>
+
+:marked
+  ## Services
+
+  ### Singletons
+  <a id="040"></a>
+  #### Style 040
+
+    - Services are singletons and should be used for sharing data and functionality.
+
+  ```typescript
+  // service  
+  import { Injectable } from 'angular2/core';
+  import { Session } from './session';
+
+  @Injectable()
+  export class HeroService {
+    getHeroes() {
+      return this._http.get('api/sessions')
+        .map((response: Response) => <Session[]>response.json().data)
+    }
+  }
+  ```
+  
+<a href="#toc">Back to top</a>
+
+:marked
+  ### Providing a Service
+  <a id="041"></a>
+  #### Style 041
+
+    - Services should be provided to the Angular 2 injector at the top-most component where they will be shared. 
```

Comment:
Maybe we should even before this section introduce the concept of component hierarchies as a separate recommendation. Everything in an Angular app is based on this single concept (the component composition, routing, di, directory structure, lazy-loading, compilation units....).

The general guidance for creating reusable components or defining providers should be that they should be created at the level of deepest common parent. This sets the visibility of that component or service instance within the application tree.